update the test script

RonEld opened the pull request:
## Description
Update `mbedtls_test.py` script to work with Python 3.7.
resolves #2653


## Status
**READY**

## Requires Backporting
Possibly, since this is in our test infrastructure, but the On Target Tests is not introduced yet in `mbedtls-2.7` ( It will be in #2075 )

## Additional comments
The CI will possibly need to change
Ran the script with the entropy on target test

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Changelog updated
- [ ] Backported

Patater commented on tests/scripts/mbedtls_test.py:
```diff
@@ -90,20 +89,24 @@ def __parse(self, data_f):
+            line = data_f.readline().strip()
             if not line:
-                continue
+                break
```
It looks like previously, if a line were empty, we'd skip it. Now, it appears we'd stop the whole loop if we run into a blank line. Is this the case? Should we use `continue`?

RonEld replied:
Yes, You are probably right. But we need to distinguish between a blank line and EOF.
Previously, the loop was over the lines in the file. Now, it's an infinite loop, and we need a break condition.
Perhaps:
```suggestion
            line = data_f.readline()
            if not line:
                break
            line = line.strip()
            if not line:
                continue	            
```

Patater commented on tests/scripts/mbedtls_test.py:
```diff
@@ -90,20 +89,24 @@ def __parse(self, data_f):
         """
-        for line in data_f:
-            line = line.strip()
+        while True:
```
Did this loop actually need to change? Doesn't `for line in file` still work in python3?

RonEld replied:
Once I had to change to `readline()` I got the error described in https://stackoverflow.com/questions/826493/mixing-files-and-loops

Patater commented on tests/scripts/mbedtls_test.py:
```diff
@@ -90,20 +89,24 @@ def __parse(self, data_f):
             # Check dependencies
             dependencies = []
-            line = data_f.next().strip()
+            line = data_f.readline().strip()
```
Can we use `line = next(data_f).strip()` in both python2.7 and python3?

RonEld replied:
I haven't tried. I'll check

Patater replied:
If we can use this, maybe we don't have to change the loop that much.

RonEld replied:
I agree, and I think it will work. I am currently working on the false positive I encountered

Patater replied:
Looks good

RonEld commented:
@Patater 
>Does this still work in python2.7?

At first I got failures, with the changes, but at the end, it succeeded.. It could be an issue of environment mixup. However, I have now noticed that I got false positive, after applying the change from https://github.com/ARMmbed/mbedtls/pull/2660#discussion_r288889691
I am working on this

RonEld commented:
@Patater I addressed your comments.

Due to the way Python3 handles encoding, I changed the data to be sent to the target as hex strings.
Unfortunately, this doesn't work on python2:
~~~
AttributeError: 'bytearray' object has no attribute 'hex'
~~~
I tried finding a common code to do same behaviour, but unfortunately, I couldn't find one that will be translated same for python3 and python2.
THe closest I could find was using [str.format](https://stackoverflow.com/questions/19210414/byte-array-to-hex-string), but then I got a differnet error, when adding `param` `char*` to `data_bytes` as a byte string.

I'll try something else

RonEld commented:
@Patater I believe I have made the script portable to both `pyhton3.7` and to `python2.7`

Patater commented on tests/scripts/mbedtls_test.py:
```diff
@@ -310,7 +309,7 @@ def run_test(self, name, function_id, dependencies, args):
         param_bytes, length = self.test_vector_to_bytes(function_id,
                                                         dependencies, args)
-        self.send_kv(length, param_bytes)
+        self.send_kv(''.join('{:02x}'.format(x) for x in length), ''.join('{:02x}'.format(x) for x in param_bytes))
```
.decode() doesn't work on both pythons?

RonEld replied:
It does, but results in different output

RonEld replied:
The problem happens when sending a hex string in the test.
The decode tries to translate it to a string, and sends wrong data
For example:
~~~
>>> data=b'\x00\x11\x22\x33\x44'
>>> print(data.decode())
"3D
~~~
The target doesn't translate this information correct.
This is on both pythons

Patater commented on tests/suites/target_test.function:
```diff
@@ -59,10 +59,43 @@ int verify_dependencies( uint8_t count, uint8_t * dep_p )
+        c = ( c - 'A' ) + 10;
+
+    byte += c ;
+    return( byte);
```
Style: funny parenthesis here

Patater commented on tests/suites/target_test.function:
```diff
@@ -59,10 +59,43 @@ int verify_dependencies( uint8_t count, uint8_t * dep_p )
+    else if( c >= 'a' && c <= 'f' )
+        c = ( c -'a' ) + 10;
+    else if( c >= 'A' && c <= 'F' )
+        c = ( c - 'A' ) + 10;
```
Should we use `strtol()` instead of implementing ourselves? Is `strtol()` portable enough (e.g. win32, etc.)? Or should we use our `hexify()` and `unhexify()` functions?

RonEld replied:
I'll check

RonEld replied:
strtol is defined in Windows, I am not sure about other platforms.
However, `strtol` receive a nul terminated string, but `c` is only a character

RonEld replied:
I used `unhexify()`

Patater commented on tests/suites/target_test.function:
```diff
@@ -59,10 +59,43 @@ int verify_dependencies( uint8_t count, uint8_t * dep_p )
+    if( c >= '0' && c <= '9' )
+        c -= '0';
+    else if( c >= 'a' && c <= 'f' )
+        c = ( c -'a' ) + 10;
```
Style: Add space after `-`.

Patater commented on tests/suites/target_test.function:
```diff
@@ -59,10 +59,43 @@ int verify_dependencies( uint8_t count, uint8_t * dep_p )
+    if( c >= '0' && c <= '9' )
+        c -= '0';
+    else if( c >= 'a' && c <= 'f' )
+        c = ( c -'a' ) + 10;
```
Style: Add space after `-`.

RonEld commented:
@Patater Thank you for the review!
I have addressed your comments

gilles-peskine-arm commented on tests/scripts/mbedtls_test.py:
```diff
@@ -79,8 +79,7 @@ def __escaped_split(inp_str, split_char):
             raise ValueError('Expected split character. Found string!')
-        out = map(split_colon_fn, re.split(r'(?<!\\)' + split_char, inp_str))
-        out = [x for x in out if x]
+        out = list(map(split_colon_fn, re.split(r'(?<!\\)' + split_char, inp_str)))
```
The code used to remove empty pieces, but no longer does. I don't understand why empty pieces would have been removed, so it may have been a bug, but since you're changing the behavior, please explain why in the commit message.

RonEld replied:
I think it's related to https://github.com/ARMmbed/mbedtls/pull/2207/files#r234792082
In addition, I used the script from https://github.com/ARMmbed/mbedtls/issues/2653#issuecomment-496234452 as reference and modified it to fit the correct versions.
I'll return this line

gilles-peskine-arm commented on tests/suites/target_test.function:
```diff
@@ -59,10 +59,29 @@ int verify_dependencies( uint8_t count, uint8_t * dep_p )
     return( DEPENDENCY_SUPPORTED );
 }
 
+/**
```
Why does the data encoding need to change? Why not fix the Python script to work with binary data correctly?

RonEld replied:
I tried it, but python 2.7 and python 3.7 behave differently with the decoding.
Specifically, when trying to generate the hex string,we get issues, as we need to supply encoding in python3, and since there are bytes that are not part of encoding, the data is wrong. 
See https://github.com/ARMmbed/mbedtls/pull/2660#discussion_r289874079 as example

gilles-peskine-arm replied:
Yes, you need to declare encodings properly. Python 2 is often a pain because a string is just a list of characters and you silently get wrong data if you pass that string to a function that expects a string in a specific encoding. Python 3 is often a pain because there are two types of strings but it isn't always clear what functions expect. But all this means is that the Python script needs to be fixed. Why would you change the communication protocol? Both Python languages are definitely capable of manipulating binary data.

RonEld replied:
yes, but they represent the binary data differently
In python2 it is just bunary data, and in python 3 it is b'....'
If we want same output for both versions, we need to change the communication.

gilles-peskine-arm replied:
They represent the data differently internally, but we don't care. If the script doesn't emit the desired format, it's a bug in the script. There's no need to change the protocol. Just fix the bug.

RonEld replied:
It's not internally. It's the way the bytes are being sent to the target

gilles-peskine-arm replied:
Exactly. If the script sends the wrong bytes to the target, that's a bug in the script.

RonEld replied:
but sending the same binary data on both python3 and python2  u8siong same code can't be done. At least not as far as I seen. This is why I changed the format of the connection

gilles-peskine-arm replied:
It can be done. I can help if you tell me exactly how to run this script so I can test what I'm doing.

gilles-peskine-arm replied:
We chatted about this and concluded that the problem is not the python version, but some evolution in greentea or one of the libraries it uses, which is now explained in the commit message.

gilles-peskine-arm commented on tests/suites/target_test.function:
```diff
@@ -71,10 +90,10 @@ int verify_dependencies( uint8_t count, uint8_t * dep_p )
-    value |= (uint8_t)greentea_getc() << 16;
-    value |= (uint8_t)greentea_getc() << 8;
-    value |= (uint8_t)greentea_getc();
+    value =  receive_byte() << 24;
```
Instead of calling `receive_byte` 4 times, how about calling `receive_hexdigit` 8 times?

RonEld replied:
Why? What's the added value?  Calling `receive_hexdigit` 8 times is equivalent to calling `greentea_getc()` 8 times, and then call `assert( unhexify( &byte, &c ) != 8 );`

RonEld commented:
@gilles-peskine-arm I addressed your comments.
I also added a comment in the commit message, about the reasoning for changing the messaging fuinctionality, as discussed offline.

In addition, running further tests, I noticed the following error in some on target tests:
~~~
ValueError: invalid literal for int() with base 10: '0x0000ffff'
~~~
I fixed this issue as well, in the last commit

gilles-peskine-arm commented on tests/scripts/mbedtls_test.py:
```diff
@@ -261,21 +261,26 @@ def test_vector_to_bytes(self, function_id, dependencies, parameters):
             if typ == 'int' or typ == 'exp':
-                i = int(param)
-                data_bytes += 'I' if typ == 'int' else 'E'
+                if re.match(r'(\d+|0x[0-9a-f]+)$',
```
Don't rewrite the standard library: `i = int(param, 0)`

RonEld replied:
Without this change, you will get the error:
~~~
`ValueError: invalid literal for int() with base 10:` that
is caused when a parameter is given in base 16. Use relevant base
when calling `int()` function.
~~~
It's either this change, or change the `generate_test_code.py` script to generate all the integers to base 10. ( some crypto tests have integers in base 16, for example [here](https://github.com/ARMmbed/mbed-crypto/blob/86268e1d302355ad8fd8e5ca5f1a7c7af8640678/tests/suites/test_suite_psa_crypto_init.data#L26))

gilles-peskine-arm replied:
Just change `i = int(param)` to `i = int(param, 0)`.

RonEld replied:
yes, thanks. I missed this part in the python documentation

RonEld commented:
@gilles-peskine-arm I amended the previous commit with your suggestion, since it's a small one line change.

RonEld commented:
CI failures happen in the following builds:
`all_sh-ubuntu-16.04-check_names`
~~~
  AR    libmbedx509.a

  AR    libmbedcrypto.a

ar: padlock.o: No such file or directory

Makefile:120: recipe for target 'libmbedcrypto.a' failed

make[2]: *** [libmbedcrypto.a] Error 1

Makefile:152: recipe for target 'libmbedcrypto.a' failed

make[1]: *** [libmbedcrypto.a] Error 2

make[1]: *** Waiting for unfinished jobs....

Makefile:19: recipe for target 'lib' failed

make: *** [lib] Error 2

+ rm list-symbols.make.log

+ mv include/mbedtls/config.h.bak include/mbedtls/config.h

+ [ -n 2 ]

+ exit 2

+ echo FAILED UNEXPECTEDLY, status=2

+ cat check-names.err

^^^^check_names: test/build: declared and exported names: tests/scripts/check-names.sh -v -> 2^^^^
~~~
`all_sh-ubuntu-16.04-test_default_cm…`
~~~
DTLS fragmenting: 3d, gnutls client, DTLS 1.0 .......................... FAIL
~~~

gilles-peskine-arm commented:
The `check_names` failure on HEAD is the race condition in build scripts that we've fixed since this PR was made. It's passing on the merge so that's ok. I don't know about the DTLS failure, but it's obviously not due to this PR.

RonEld commented:
Isn't the DTLS error the known 3d gnutls issue we are facing?

RonEld commented:
Backported to `mbedtls-2.16` in #2744 
The on target tests do not exist yet in 2.17, so no need to backport there, but possibly cherry-picking to #2075
